In [5]:
# checking which environment it ran on
import sys
print(sys.executable)

/home/aviraljoshi/mini-logistic-regression/.venv/bin/python3


In [6]:
# importing numpy as np
import numpy as np
print(np.__version__)

2.5.3


In [7]:
heights_cm = np.array([170, 155, 182])
print(heights_cm)
print(type(heights_cm))

[170 155 182]
<class 'numpy.ndarray'>


In [8]:
print(heights_cm.shape)
print(type(heights_cm.shape))

(3,)
<class 'tuple'>


In [9]:
print(heights_cm.ndim)
print(heights_cm.dtype)

1
int64


In [10]:
weights_kg = np.array([65.5, 50.0, 80.2])
print(weights_kg.ndim)
print(weights_kg.dtype)

1
float64


In [13]:
# Indexing and slicing a 1D array
print(heights_cm[0])
print(heights_cm[-1])
print(heights_cm[0:2])

# slicing also returns a ndarray
print(type(heights_cm[0:2]))

170
182
[170 155]
<class 'numpy.ndarray'>


In [14]:
# Loops vs vectorization
numbers = list(range(1_000_000))
numbers_array = np.array(numbers)

%timeit [n * 2 for n in numbers]
%timeit numbers_array * 2

37.5 ms ± 2.68 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
848 μs ± 75.1 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


In [16]:
#Vectors and matrices (the concept)
X = np.array([
    [14.1, 20.3, 92.0],
    [11.4, 15.7, 73.2],
    [19.8, 22.1, 130.5],
    [12.5, 18.9, 80.1],
])

print(X)
print(X.shape)
print(X.ndim)
print(X.dtype)

[[ 14.1  20.3  92. ]
 [ 11.4  15.7  73.2]
 [ 19.8  22.1 130.5]
 [ 12.5  18.9  80.1]]
(4, 3)
2
float64


In [25]:
# Indexing a matrix (rows, columns, single values)

print(X[0])
print(X[:, 0])
print(X[2, 2])


print(X[0].shape, X[:, 0].shape)

[14.1 20.3 92. ]
[14.1 11.4 19.8 12.5]
130.5
[[14.1]
 [11.4]]
[20.3 15.7]
(3,) (4,) (2, 1)


In [26]:
print(X[:2, :1])
print(X[:2, 1])

print(X[:2, :1].shape, X[:2, 1].shape)

[[14.1]
 [11.4]]
[20.3 15.7]
(2, 1) (2,)


In [28]:
# Broadcasting (the concept)

marks = np.array([
    [10, 20, 30],
    [40, 50, 60],
])
bonus = np.array([1, 2, 3])

print(marks.shape, bonus.shape)
print(marks - bonus)
print((marks - bonus).shape)

(2, 3) (3,)
[[ 9 18 27]
 [39 48 57]]
(2, 3)


### NumPy Broadcasting Rules Reference

Broadcasting allows arithmetic operations on arrays of different shapes. For an operation to work, the dimensions of the two arrays must be compatible. 

**The Alignment Rule:** Always align shapes starting from the **right side** (trailing dimensions first) [2.1]. If one array has fewer dimensions, NumPy automatically prepends `1`s to its shape to match the length of the larger array [2.1].

| Shape A | Shape B | Expanded B | Rows Comparison | Columns Comparison | Status & Behavior |
| :--- | :--- | :--- | :--- | :--- | :--- |
| `(2, 3)` | `(2, 3)` | `(2, 3)` | Equal ✓ | Equal ✓ | **Works** (Standard element-by-element operation) |
| `(2, 3)` | `(1, 3)` | `(1, 3)` | 1 → Stretch ✓ | Equal ✓ | **Works** (Shape B stretches rows to match A) |
| `(2, 3)` | `(2, 1)` | `(2, 1)` | Equal ✓ | 1 → Stretch ✓ | **Works** (Shape B stretches columns to match A) |
| `(2, 3)` | `(1, 1)` | `(1, 1)` | 1 → Stretch ✓ | 1 → Stretch ✓ | **Works** (Acts like adding a single scalar) |
| `(2, 3)` | `(3,)` | `(1, 3)` | 1 → Stretch ✓ | Equal ✓ | **Works** ✓ (B aligns right, stretches across rows) [2.1] |
| `(2, 3)` | `(2,)` | `(1, 2)` | Equal ✓ | 3 vs 2 ✗ | ❌ **Error** (Trailing dimensions `3` and `2` mismatch) [2.1] |
| `(2, 3)` | `(2, 2)` | `(2, 2)` | Equal ✓ | 3 vs 2 ✗ | ❌ **Error** (Dimension mismatch; cannot stretch) |

---

### Core Takeaways

* **Right-to-Left Alignment:** Before comparing, NumPy pads the smaller shape on the *left* with `1`s [2.1]. This is why `(3,)` becomes `(1, 3)` (compatible) [2.1], but `(2,)` becomes `(1, 2)` (incompatible) [2.1].
* **The Rule of 1:** A dimension of size `1` acts like a rubber band. NumPy will automatically stretch/copy it to match the larger shape along that axis without using extra memory.
* **The Mismatch Rule:** If two aligned dimensions are different and neither is `1`, broadcasting fails, throwing a `ValueError` [2.1].

### Quick Fix for the `(2,)` Mismatch
To make a `(2,)` array work with a `(2, 3)` array, you must manually transform its shape from a row vector into a column vector `(2, 1)` [2.1] so that the trailing dimension is a stretchable `1`:
```python
# Convert shape (2,) to (2, 1) to force broadcasting across columns
array_B[:, np.newaxis]  # or array_B.reshape(-1, 1)
```


In [30]:
# Transpose (the concept)

A = np.array([
    [1, 2, 3],
    [4, 5, 6],
])

print(A.T)
print(A.shape, A.T.shape)
print(A[1,2], A.T[2,1])
print(np.array_equal(A.T.T, A))

[[1 4]
 [2 5]
 [3 6]]
(2, 3) (3, 2)
6 6
True


In [31]:
# The dot product (the concept, by hand)

x = np.array([2, 3, 1])
w = np.array([0.5, -1, 2])

print(x * w)
print(np.sum(x * w))
print(np.dot(x, w))
print(x @ w)

[ 1. -3.  2.]
0.0
0.0
0.0


In [33]:
X_small = np.array([
    [2, 3, 1],
    [4, 0, 1],
])

scores = X_small @ w
print(scores)
print(X_small.shape, w.shape, scores.shape)

print(X_small[0]@w, X_small[1]@w)

[0. 4.]
(2, 3) (3,) (2,)
0.0 4.0


In [34]:
y_small = np.array([0, 1])

print((scores - y_small).shape)

scores_column = scores.reshape(2, 1)
print(scores_column.shape)
print(scores_column - y_small)

(2,)
(2, 1)
[[ 0. -1.]
 [ 4.  3.]]


By the broadcasting rule: (2, 1) and (2,) lined up from the right means 1 vs 2, so the column gets stretched across; the missing left number is filled, so the row gets copied down. Both stretch, giving a 2 × 2 grid. It subtracted every label from every score, including patient 0's score minus patient 1's label, which is meaningless. And no error appeared.

In [35]:
b = -1.0

z = X_small @ w + b
print(z)
print(z.shape)

[-1.  3.]
(2,)


In [38]:
# np.mean and the axis argument

marks = np.array([[10, 20, 30],
                 [40, 50, 60]])
print(np.mean(marks, axis = 0))
print(np.mean(marks, axis = 1))

[25. 35. 45.]
[20. 50.]


#### The Number `e`

The number **\(e\)** is a special mathematical constant, much like \(\pi\), roughly equal to **\(2.718\)**. It appears naturally in patterns of growth and probability, which is exactly why the formula for **logistic regression** is built on top of it. 

The expression **\(e^z\)** means *"\(e\) raised to the power of \(z\)"*.

> **\(e^z\) is always positive.** 
> Even for highly negative values of \(z\), the result only gets closer and closer to \(0\). It never actually reaches \(0\), and it never drops below it into negative numbers.

> **It grows extremely fast.** 
> Each \(+1\) added to \(z\) multiplies the previous result by roughly \(2.7\). By the time \(z = 10\), the value has already skyrocketed to **\(\approx 22,026\)**.


In [39]:
# np.exp

print(np.e)
print(np.exp(z))

z_values = np.array([-2.0, -1.0, 0.0, 1.0, 2.0])
print(np.exp(z_values))

2.718281828459045
[ 0.36787944 20.08553692]
[0.13533528 0.36787944 1.         2.71828183 7.3890561 ]


## When `np.exp` overflows

The exponential function grows **very quickly**:

\[
e^z
\]

Each time we increase `z` by `1`, the value of \(e^z\) is multiplied by approximately **2.718**.

For example:

- \(e^1 \approx 2.718\)
- \(e^2 \approx 7.389\)
- \(e^{10} \approx 22,026\)
- \(e^{100} \approx 2.69 \times 10^{43}\)

### What is overflow?

A NumPy `float64` can store values only up to approximately:

\[
1.8 \times 10^{308}
\]

If `np.exp(z)` becomes larger than this limit, there is no room to store the result.

This is called **overflow**.

```python
np.exp(1000)
# inf

In [40]:
big_z = np.array([100.0, 700.0, 710.0, 1000.0])
print(np.exp(big_z))

print(np.exp(-1000.0))

[2.68811714e+043 1.01423205e+304             inf             inf]
0.0


/tmp/ipykernel_19785/2443187023.py:2: RuntimeWarning: overflow encountered in exp
  print(np.exp(big_z))


## The logarithm and `np.log`

### What is a logarithm?

A **logarithm** undoes a power.

The **natural logarithm**, written as `ln`, undoes exponentiation with the base \(e\).

In other words:

\[
\ln(x) = \text{"e to what power gives } x\text{"?"}
\]

For example:

\[
e^2 \approx 7.389
\]

Therefore:

\[
\ln(7.389) \approx 2
\]

So:

```python
np.log(7.389)
# ≈ 2

In [41]:
x_values = np.array([np.e, 1.0, 0.5, 0.01, 0.0, -1.0])
print(np.log(x_values))

print(np.log(np.exp(3.0)))

[ 1.          0.         -0.69314718 -4.60517019        -inf         nan]
3.0


/tmp/ipykernel_19785/2555687726.py:2: RuntimeWarning: divide by zero encountered in log
  print(np.log(x_values))
/tmp/ipykernel_19785/2555687726.py:2: RuntimeWarning: invalid value encountered in log
  print(np.log(x_values))


## Clipping

### The concept

**Clipping** means forcing every number to stay between a **lower limit** and an **upper limit**.

The rule is simple:

- If a value is **below the lower limit** → raise it to the lower limit.
- If a value is **above the upper limit** → lower it to the upper limit.
- If a value is **between the limits** → leave it unchanged.

### Example

Suppose we want every value to stay between `0` and `10`:

| Original | Clipped |
|---:|---:|
| `-5` | `0` |
| `0` | `0` |
| `3` | `3` |
| `10` | `10` |
| `15` | `10` |

So clipping doesn't change values that are already inside the allowed range.

### In NumPy

NumPy provides `np.clip()` for this:

```python
np.clip(x, lower_limit, upper_limit)

In [42]:
eps = 1e-15
# eps = 1e-15 is 0.000000000000001. eps is short for epsilon

probs = np.array([0.0, 0.3, 1.0])

safe_probs = np.clip(probs, eps, 1 - eps)
print(safe_probs)
print(np.log(safe_probs))

[1.e-15 3.e-01 1.e+00]
[-3.45387764e+01 -1.20397280e+00 -9.99200722e-16]


In [43]:
# np.maximum vs np.max

a = np.array([-3.0, 0.0, 5.0])

print(np.max(a))
print(np.maximum(a, 0))
print(np.maximum(a, 0).shape)

5.0
[0. 0. 5.]
(3,)
